ledger/src/slot_stats.rs

```diff
@@ -0,0 +1,90 @@
+use {
+    crate::blockstore_meta::SlotMeta, bitflags::bitflags, lru::LruCache, solana_sdk::clock::Slot,
+};
+
+const SLOTS_STATS_CACHE_CAPACITY: usize = 300;
+
+macro_rules! get_mut_entry (
+    ($cache:expr, $key:expr) => (
+        match $cache.get_mut(&$key) {
+            Some(entry) => entry,
+            None => {
+                $cache.put($key, SlotStats::default());
+                $cache.get_mut(&$key).unwrap()
+            }
+        }
+    );
+);
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) enum ShredSource {
+    Turbine,
+    Repaired,
+    Recovered,
+}
+
+bitflags! {
+    #[derive(Default)]
+    struct SlotFlags: u8 {
+        const DEAD   = 0b00000001;
+        const FULL   = 0b00000010;
+        const ROOTED = 0b00000100;
+    }
+}
+
+#[derive(Default)]
+struct SlotStats {
+    flags: SlotFlags,
+    num_repaired: usize,
+    num_recovered: usize,
+}
+
+pub(crate) struct SlotsStats(LruCache<Slot, SlotStats>);
+
+impl Default for SlotsStats {
+    fn default() -> Self {
+        // LruCache::unbounded because capacity is enforced manually.
+        Self(LruCache::unbounded())
+    }
+}
+
+impl SlotsStats {
+    pub(crate) fn add_shred(&mut self, slot: Slot, source: ShredSource) {
+        let entry = get_mut_entry!(self.0, slot);
+        match source {
+            ShredSource::Turbine => (),
+            ShredSource::Repaired => entry.num_repaired += 1,
+            ShredSource::Recovered => entry.num_recovered += 1,
+        }
+        self.maybe_evict_cache();
+    }
+
+    pub(crate) fn set_full(&mut self, slot_meta: &SlotMeta) {
+        let total_time_ms =
+            solana_sdk::timing::timestamp().saturating_sub(slot_meta.first_shred_timestamp);
+        let last_index = slot_meta
+            .last_index
+            .and_then(|ix| i64::try_from(ix).ok())
+            .unwrap_or(-1);
+        let entry = get_mut_entry!(self.0, slot_meta.slot);
+        if !entry.flags.contains(SlotFlags::FULL) {
+            datapoint_info!(
+                "shred_insert_is_full",
+                ("total_time_ms", total_time_ms, i64),
+                ("slot", slot_meta.slot, i64),
+                ("last_index", last_index, i64),
+                ("num_repaired", entry.num_repaired, i64),
+                ("num_recovered", entry.num_recovered, i64),
```

Comment:
If this slot was evicted from the cache and re-inserted later it's possible these two metrics could be incorrect right? In these cases should we mark the entry so that we report some default value like -1 for these entries to distinguish them from the other entries?

Reply:
> I am expecting we do not receive shreds from turbine after the slot is evicted from the cache here.

So this will happen if we have `>300` slots that received a shred but aren't full right? Referencing the discussion here: https://github.com/solana-labs/solana/pull/23872#discussion_r834338421

>It is only a problem if the lru cache capacity is not enough to cover all the slots that we are actively >receiving shreds for. i.e. it should not matter how far behind the node is. The node receives shreds from >turbine for the slots close to the tip + shreds from repair for the slots it chooses to repair. 

I think this is more likely the further you are behind the tip because there are more slots near the tip that you will not choose to repair fully (repair always tries to fully repair slots connected to the root first)

> If still that scenario happens, in the post-processing of metrics, if there are multiple rows of data for a slot, we can pick the one with the earliest timestamp and discard the rest.

Hmmm, we don't report currently on eviction right, is that this TODO here: https://github.com/solana-labs/solana/blob/81d88f0ebb957e50e43fafa3adfc6cdb38eb97a9/ledger/src/slot_stats.rs#L87

> We can iterate and revise this if lru eviction policy and the capacity of 300 was not enough

Currently what would be the best way to tell from the metrics that this is happening? It seems we would report a datapoint whether or not the slots were being evicted from the cache.

What if we add a "num_turbine" here: https://github.com/solana-labs/solana/blob/81d88f0ebb957e50e43fafa3adfc6cdb38eb97a9/ledger/src/slot_stats.rs#L55.

Then we know if `num_turbine` + `num_repaired` + `num_recovered` != last_index + 1 it must have been evicted

Reply:
I was logging whenever an entry is removed from the cache and included bool values for rooted/dead/evicted/pruned so these entries could be filtered by a `WHERE` clause.

Reply:
@carllin @jbiseda 

* I think at this point we need some very basic and simple metrics.
* It also has to be _very lightweight_ not to slow down shred ingestion just to emit metrics.
* Once we have those metrics we can do some post analysis on the data and check what is missing and add more metrics if needed.